feat(74/rln-stealth-commitments): add initial tech writeup (#658)
vacp2p/rfc: content/docs/rfcs/73/README.md

---
slug: 74
title: 74/STATUS-COMMUNITY-STEALTH-COMMITMENTS
name: Stealth commitment usage for Status community members
status: raw
category: Proof of Concept
editor: Aaryamann Challani <aaryamann@status.im>
contributors:
- Jimmy Debe <jimmy@status.im> 
---

# Abstract

This document describes the usage of stealth commitments to add prospective members of a [Status Community](/spec/56) to the network-governed [RLN](/spec/32) membership set.

# Motivation

When [RLN](/spec/32) is enforced in [Waku](/spec/10), 

> all users of the Status application will be required to register to the membership set.

Reviewer: We can keep the spec more general:

> Any users willing to send message must have a RLN membership registered on the contract. Forcing user to do an onchain transaction to use an application is an onboarding friction that projects may not be willing to accept.

> This document describes a privacy-preserving mechanism in which a counterparty MAY register a commitment that only the party can generate a valid private key for.

Reviewer: If you want to add more context:

> said counterparty may be a different type of end user or developers of a given app.

# Theory

This specification describes an implementation of the [Stealth Address](https://eips.ethereum.org/EIPS/eip-5564) scheme, 
tailored to the curve that is used in the [RLN](/spec/32) protocol.

> This can be used in two places in the Status Application:

Reviewer: Let's keep it generalized:

> 1. It can enable special users such as developer of a project or, in the context of Status, Community Owner to...

1. For community owners to aggregate prospective members of their community, and add them to the [RLN](/spec/32) membership set in a batch, or
2. For users of the application to register someone else to the [RLN](/spec/32) membership set, without revealing their identity.

> This is useful in when the prospective user does not have access to funds on the network that [RLN](/spec/32) is deployed on.

Reviewer: ```suggestion
This is useful in when the prospective user does not have access to funds on the network that [RLN](/spec/32) is deployed on or are unwilling to use such found to try an applicaiton.
```

---
slug: 74
title: 74/RLN-STEALTH-COMMITMENTS
name: RLN Stealth Commitment Usage
status: raw
category: Standards Track
description:
editor: Aaryamann Challani <aaryamann@status.im>
contributors:
- Jimmy Debe <jimmy@status.im>
---

## Abstract

This specification describes the usage of stealth commitments to add prospective users to a network-governed [32/RLN-V1](/spec/32) membership set.

## Motivation

When [32/RLN-V1](/spec/32) is enforced in [10/Waku2](/spec/10), 
all users are required to register to a membership set.
The membership set will store user identities allowing the secure interaction within an application.
Forcing a user to do an on-chain transaction to join a membership set is an onboarding friction, 
and some projects may be opposed to this method.
To improve the user experience,

> stealth commitments can be used by a counterparty to register identites on the user's behalf,

Reviewer: ```suggestion
stealth commitments can be used by a counterparty to register identities on the user's behalf,
```

while maintaining the user's anonymity.

This document specifies a privacy-preserving mechanism,
allowing a counterparty to utilize [32/RLN-V1](/spec/32) to register an `identityCommitment` on-chain.
Counterparties will be able to register members to a RLN membership set without exposing the user's private keys.

## Background
The [32/RLN-V1](/spec/32) protocol, 
consists of a smart contract that stores a `idenitityCommitment` in a membership set.
In order for a user to join the membership set,
the user is required to make a transaction on the blockchain.
A set of public keys are used to compute a stealth address,

> as desribed in [ERC-5564](https://eips.ethereum.org/EIPS/eip-5564).

Reviewer: ```suggestion
as described in [ERC-5564](https://eips.ethereum.org/EIPS/eip-5564).
```

This specification is an implementation of the [ERC-5564](https://eips.ethereum.org/EIPS/eip-5564) scheme, 
tailored to the curve that is used in the [32/RLN-V1](/spec/32) protocol.

This can be used in couple of ways in applications:
1. Applications can add users to the [32/RLN-V1](/spec/32) membership set in a batch.
2. Users of the application can register other users to the [32/RLN-V1](/spec/32) membership set.

This is useful when the prospective user does not have access to funds on the network that [32/RLN-V1](/spec/32) is deployed on.

## Wire Format Specification 

> The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, 

Reviewer: I don't think this needs to be repeated, I believe it's already part of 1/COSS

Reviewer 2: The IETF puts this in each document.
We should be consistent though, having this in each RFC or in none. No strong opinion here.
Also it should not be in the "Wire Format Specification ", you can check an IETF RFC for reference.

“NOT RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

The wire formats are encoded using protobuf.
The wire formats that SHOULD be used by implementors include two message types, 
`RequestToJoin` and `JoinResponse`.

### Request to Join
The `RequestToJoin` message MUST be used by users joining a membership set with assistance from the application or 
other users.
The message should include `spending_public_key` and 

> a `viewing_public_key` that is generated by the member’s public key.

Reviewer: ```suggestion
The message MUST include `spending_public_key` and 
a `viewing_public_key` that is generated from the member’s public key.
```

is "member" the user here?

The protobuf message description is:

```protobuf

syntax = "proto3";

message RequestToJoin {
  // The Lamport timestamp of the request  
  uint64 clock = 1;

  // The ENS name of the requester
  string ens_name = 2;

  // The chat id requested to join
  string pubsub_topic = 3;

  // The public key of the application
  bytes app_key = 4;

  // The display name of the requester
  string display_name = 5;

  // The spending public key of the requester
  bytes spending_public_key = 6;

  // The viewing public key of the requester
  bytes viewing_public_key = 7;
}

> 

Reviewer: ```suggestion
message RequestToJoin {
  // The spending public key of the requester
  bytes spending_public_key = 6;

  // The viewing public key of the requester
  bytes viewing_public_key = 7;
}

```

None of the other fields seem to be necessary

```

The `spending_public_key` and
`viewing_public_key` MUST be included in the message.
Both attributes create a stealth meta-address,
as described in [ERC-5564](https://eips.ethereum.org/EIPS/eip-5564).
- both SHOULD be a 32-byte hash.

> - both SHOULD be derived from the community member's public key.

Reviewer: ```suggestion
- both SHOULD be derived from the community member's public key.
```
This is Status terminology.

I assume you mean "user public key".

Also, there is a pre-requisite that users generate a key pair. Would be good to have that in the spec.

### Generate Stealth Address
The application or user SHOULD generate a stealth address after the message is received.

> This address MUST be used to publish the requester's [32/RLN-V1](/spec/32) `identitycommitment` for the corresponding `app_key` memebership set.

Reviewer: what is an `app_key`? Would be good to define that as pre-requisite.

Reviewer 3: Changed to application membership set.

- The receiver MUST generate an `ephemeral_public_key` and `stealth_commitment`.
This will be used to check the stealth address used to update the memebership set,
and the user MUST be able to check with their `viewing_public_key`.

### Response for Request to Join
The `JoinResponse` message MUST be used by applications/users willing to service the requester.

```protobuf

syntax = "proto3";

message JoinResponse {

>   // The Lamport timestamp of the request
>   uint64 clock = 1;
> 
>   // The community description
>   CommunityDescription community = 2;
> 
>   // If the request was accepted
>   bool accepted = 3;
> 
>   // The grant offered by the community
>   bytes grant = 4;
> 
>   // The community public key
>   bytes community_id = 5;
> 
>   // The stealth commitment for the requester
>   bytes stealth_commitment = 6;
> 
>   // The ephemeral public key used to generate the commitment
>   bytes ephemeral_public_key = 7;
> }

Reviewer: ```suggestion
  // If the request was accepted
  bool accepted = 3;

  // The grant offered by the community
  bytes grant = 4;

  // The stealth commitment for the requester
  bytes stealth_commitment = 6;

  // The ephemeral public key used to generate the commitment
  bytes ephemeral_public_key = 7;
}
```

Removing everything that seems to be Status community related.

What is `grant` ?

for `accepted`, maybe best to have a `status_code` similar to other protocols: https://github.com/vacp2p/rfc/blob/84638042ce10b8416918056c93ca32bbd13556c9/content/docs/rfcs/12/README.md?plain=1#L97

- The community owner MUST generate a `ephemeral_public_key` and `stealth_commitment`.
- The [32/RLN-V1](/spec/32) community membership set MUST be updated by the community owner.

### Response for Request to Join
The `CommunityRequestToJoinResponse` message MUST be used by community owners.

```protobuf

syntax = "proto3";

message CommunityRequestToJoinResponse {
  // The Lamport timestamp of the request
  uint64 clock = 1;

  // The community description
  CommunityDescription community = 2;

  // If the request was accepted
  bool accepted = 3;

  // The grant offered by the community
  bytes grant = 4;

  // The community public key
  bytes community_id = 5;

  // The stealth commitment for the requester
  bytes stealth_commitment = 6;

  // The ephemeral public key used to generate the commitment
  bytes ephemeral_public_key = 7;
}

```

- The `stealth_commitment` and `ephemeral_public_key` MUST be used to generate a private key, 
the `spendingKey`.
- The community owner SHOULD be able to check `stealth_commitment` belongs to them on-chain.
- The `spendingKey` MUST be used to generate valid [32/RLN-V1](/spec/32) proofs to send messages.

## Implementation Suggestions

> An implementation of the Stealth Address scheme is available in the [erc-5564-bn254](https://github.com/rymnc/erc-5564-bn254) repository,

Reviewer 2: We should update this link in a stable. We could open a repo for this under Vac org.
Generally, this section should rather give tips on how to potentially implement specific features and not link an existing impl. (No hard requirement, still, OK to link a reference impl as it, too, serves as a "suggestion")
